api/v1alpha1/serviceintentions_types.go

```diff
@@ -296,11 +334,73 @@ func (in IntentionPermissions) validate(path *field.Path) field.ErrorList {
 
 func (in *IntentionHTTPPermission) validate(path *field.Path) field.ErrorList {
 	var errs field.ErrorList
-	if invalidPathPrefix(in.PathPrefix) {
-		errs = append(errs, field.Invalid(path.Child("pathPrefix"), in.PathPrefix, `must begin with a '/'`))
+	pathParts := 0
+	if in.PathRegex != "" {
+		pathParts++
+	}
+	if in.PathPrefix != "" {
+		pathParts++
+		if invalidPathPrefix(in.PathPrefix) {
+			errs = append(errs, field.Invalid(path.Child("pathPrefix"), in.PathPrefix, `must begin with a '/'`))
+		}
+	}
+	if in.PathExact != "" {
+		pathParts++
+		if invalidPathPrefix(in.PathExact) {
+			errs = append(errs, field.Invalid(path.Child("pathExact"), in.PathExact, `must begin with a '/'`))
+		}
+	}
+	if pathParts > 1 {
+		asJSON, _ := json.Marshal(in)
+		errs = append(errs, field.Invalid(path, string(asJSON), `At most only one of pathExact, pathPrefix, or pathRegex may be configured.`))
 	}
-	if invalidPathPrefix(in.PathExact) {
-		errs = append(errs, field.Invalid(path.Child("pathExact"), in.PathExact, `must begin with a '/'`))
+
+	found := make(map[string]struct{})
+	for i, method := range in.Methods {
+		methods := []string{
+			http.MethodGet,
+			http.MethodHead,
+			http.MethodPost,
+			http.MethodPut,
+			http.MethodPatch,
+			http.MethodDelete,
+			http.MethodConnect,
+			http.MethodOptions,
+			http.MethodTrace,
+		}
+		if !sliceContains(methods, method) {
+			errs = append(errs, field.Invalid(path.Child("methods").Index(i), method, notInSliceMessage(methods)))
+		}
+		if _, ok := found[method]; ok {
+			errs = append(errs, field.Invalid(path.Child("methods").Index(i), method, `Method contained more than once.`))
+		}
+		found[method] = struct{}{}
+	}
+	errs = append(errs, in.Header.validate(path.Child("header"))...)
+	return errs
+}
+func (in IntentionHTTPHeaderPermissions) validate(path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	for i, permission := range in {
+		hdrParts := 0
+		if permission.Present {
+			hdrParts++
+		}
+		if permission.Exact != "" {
+			hdrParts++
+		}
+		if permission.Regex != "" {
+			hdrParts++
+		}
+		if permission.Prefix != "" {
+			hdrParts++
+		}
+		if permission.Suffix != "" {
+			hdrParts++
+		}
+		if hdrParts != 1 {
+			errs = append(errs, field.Invalid(path.Index(i), in[i], `At most only one of exact, prefix, suffix, regex, or present may be configured.`))
+		}
```

Comment:
`in[i]` should be json encoded here.

Unrelated and not a blocker, I think a helper:
```go
func numNotEmpty(ss ...string) int {
	count := 0
	for _, s := range ss {
		if s != "" {
			count++
		}
	}
	return count
}
```

Would clean this up even further.